crates/completion/src/render/pattern.rs

```diff
@@ -0,0 +1,146 @@
+//! Renderer for patterns.
+
+use hir::{db::HirDatabase, HasAttrs, HasVisibility, Name, StructKind};
+use itertools::Itertools;
+
+use crate::{
+    config::SnippetCap, item::CompletionKind, render::RenderContext, CompletionItem,
+    CompletionItemKind,
+};
+
+pub(crate) fn render_struct_pat(
+    ctx: RenderContext<'_>,
+    strukt: hir::Struct,
+    local_name: Option<Name>,
+) -> Option<CompletionItem> {
+    let _p = profile::span("render_struct_pat");
+
+    let module = ctx.completion.scope.module()?;
+    let fields = strukt.fields(ctx.db());
+    let n_fields = fields.len();
+    let fields = fields
+        .into_iter()
+        .filter(|field| field.is_visible_from(ctx.db(), module))
+        .collect::<Vec<_>>();
+
+    if fields.is_empty() {
+        // Matching a struct without matching its fields is pointless, unlike matching a Variant without its fields
+        return None;
+    }
+    let fields_omitted =
+        n_fields - fields.len() > 0 || strukt.attrs(ctx.db()).by_key("non_exhaustive").exists();
+
+    let name = local_name.unwrap_or_else(|| strukt.name(ctx.db())).to_string();
+    let pat = render_pat(&ctx, &name, strukt.kind(ctx.db()), &fields, fields_omitted)?;
+
+    let mut completion = CompletionItem::new(CompletionKind::Snippet, ctx.source_range(), name)
+        .kind(CompletionItemKind::Binding)
+        .set_documentation(ctx.docs(strukt))
+        .set_deprecated(ctx.is_deprecated(strukt))
+        .detail(&pat);
+    if let Some(snippet_cap) = ctx.snippet_cap() {
+        completion = completion.insert_snippet(snippet_cap, pat);
+    } else {
+        completion = completion.insert_text(pat);
+    }
```

Comment:
```suggestion
    completion = if let Some(snippet_cap) = ctx.snippet_cap() {
        completion.insert_snippet(snippet_cap, pat);
    } else {
        completion.insert_text(pat);
    }
```

we need an :bulb: for this